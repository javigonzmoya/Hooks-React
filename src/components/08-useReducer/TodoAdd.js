import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
  const [values, handleInputChange, reset] = useForm({ desciption: '' });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (values.desciption.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: values.desciption,
      done: false,
    };

    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Agregar TODO</h4>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="desciption"
          className="form-control"
          placeholder="Hacer.."
          autoComplete="off"
          onChange={handleInputChange}
          value={values.desciption}
        />
        <div className="d-grid gap-2">
          <button className="btn btn-outline-primary mt-1 " type="submit">
            Agregar
          </button>
        </div>
      </form>
    </>
  );
};
