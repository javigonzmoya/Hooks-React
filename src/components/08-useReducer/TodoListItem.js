import React from 'react';

export const TodoListItem = ({ todo, index, handleRemove, handleToggle }) => {
  const { id, desc, done } = todo;
  return (
    <li className="list-group-item">
      <p
        className={`${done ? 'complete' : ''} text-center`}
        onClick={() => handleToggle(id)}
      >
        {index + 1}.{desc}
      </p>
      <button className="btn btn-danger" onClick={() => handleRemove(id)}>
        borrar
      </button>
    </li>
  );
};
