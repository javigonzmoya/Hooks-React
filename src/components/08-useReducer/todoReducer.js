export const todoReducer = (state = [], action) => {
  switch (action.type) {
    case 'add':
      return [...state, action.payload];
    case 'remove':
      return state.filter((s) => s.id !== action.payload);
    case 'toggle':
      return state.map((s) =>
        s.id === action.payload ? { ...s, done: !s.done } : s
      );
    case 'toggle-old':
      return state.map((s) => {
        if (s.id === action.payload) {
          return {
            ...s,
            done: !s.done,
          };
        }
        return s;
      });
    default:
      return state;
  }
};
