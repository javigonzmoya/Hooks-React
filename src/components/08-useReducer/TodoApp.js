import React, { useEffect, useReducer } from 'react';
import { useForm } from '../../hooks/useForm';
import { todoReducer } from './todoReducer';

import './styles.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);
  const [values, handleInputChange, reset] = useForm({ desciption: '' });

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleSubmit = (e) => {
    e.preventDefault();

    if (values.desciption.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: values.desciption,
      done: false,
    };

    const action = {
      type: 'add',
      payload: newTodo,
    };
    dispatch(action);
    reset();
  };

  const handleRemove = (todoId) => {
    const action = {
      type: 'remove',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId,
    };
    dispatch(action);
  };
  return (
    <div>
      <h1>TodoApp: ( {todos.length} )</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <ul className="list-group list-group-flush">
            {todos.map(({ id, desc, done }, i) => (
              <li key={id} className="list-group-item">
                <p
                  className={`${done ? 'complete' : ''} text-center`}
                  onClick={() => handleToggle(id)}
                >
                  {i + 1}.{desc}
                </p>
                <button
                  className="btn btn-danger"
                  onClick={() => handleRemove(id)}
                >
                  borrar
                </button>
              </li>
            ))}
          </ul>
        </div>

        <div className="col-5">
          <h4>Agregar TODO</h4>
          <form onSubmit={handleSubmit}>
            <input
              type="text"
              name="desciption"
              className="form-control"
              placeholder="Hacer.."
              autoComplete="off"
              onChange={handleInputChange}
              value={values.desciption}
            />
            <div className="d-grid gap-2">
              <button className="btn btn-outline-primary mt-1 " type="submit">
                Agregar
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
