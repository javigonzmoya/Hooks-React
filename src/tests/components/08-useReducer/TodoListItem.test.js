import { shallow } from 'enzyme';
import React from 'react';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas sobre <TodoListItem/>', () => {
  //{ todo, index, handleRemove, handleToggle }
  const handleRemove = jest.fn();
  const handleToggle = jest.fn();
  const wrapper = shallow(
    <TodoListItem
      todo={demoTodos[0]}
      index={0}
      handleRemove={handleRemove}
      handleToggle={handleToggle}
    />
  );

  test('Debe de mostar el componente Correctamente ', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe de llamar la funcion de handleRemove ', () => {
    const button = wrapper.find('button');
    button.simulate('click');
    expect(handleRemove).toHaveBeenCalledTimes(1);
    expect(handleRemove).toHaveBeenCalledWith(demoTodos[0].id);
  });

  test('Debe de llamar la funcion de handleToggle ', () => {
    const button = wrapper.find('p');
    button.simulate('click');
    expect(handleToggle).toHaveBeenCalledTimes(1);
    expect(handleToggle).toHaveBeenCalledWith(demoTodos[0].id);
  });

  test('Debe de mostrar el texto correctamente ', () => {
    const p = wrapper.find('p');
    expect(p.text().trim()).toBe(`${1}.${demoTodos[0].desc}`);
    expect(p.props().className.trim()).toBe(`text-center`);
  });

  test('Debe de mostrar complete si lo esta ', () => {
    const todo = demoTodos[0];
    todo.done = true;
    const wrapper = shallow(
      <TodoListItem
        todo={todo}
        index={0}
        handleRemove={handleRemove}
        handleToggle={handleToggle}
      />
    );

    const p = wrapper.find('p');
    expect(p.props().className.trim()).toBe(`complete text-center`);
    expect(p.hasClass('complete')).toBe(true); //otra forma de hacerlo
  });
});
