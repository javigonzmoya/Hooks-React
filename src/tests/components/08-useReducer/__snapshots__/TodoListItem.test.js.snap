// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <TodoListItem/> Debe de mostar el componente Correctamente  1`] = `
<li
  className="list-group-item"
>
  <p
    className=" text-center"
    onClick={[Function]}
  >
    1
    .
    Aprender react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    borrar
  </button>
</li>
`;
